### PR TITLE
Implement ContainerHash: a simple hashing structure for containers

### PR DESCRIPTION
**Pull request by rodsan0**
ContainerHash utilizes hash_combine to assign a unique hash to a container, allowing their use in sets and unordered_map keys.

### CODE REVIEW

File: source/tools/hash_utils.h [hunk 1]
```diff
@@ -24,9 +26,22 @@ namespace emp {
...
   constexpr inline std::size_t hash_combine(std::size_t hash1, std::size_t hash2)
   {
-    return hash1 ^ (hash2 * 0x9e3779b9 + (hash1 << 6) + (hash1 >> 2));
+    return hash1 ^ (hash2 + 0x9e3779b9 + (hash1 << 6) + (hash1 >> 2));
   }
 
+  template <typename Container, size_t Seed = 0>
+  struct ContainerHash
+  {
+    size_t operator()(const Container& v) const {
```
Comment (mmore500) [inline source/tools/hash_utils.h:35]: doc string?

File: source/tools/hash_utils.h [hunk 2]
```diff
@@ -24,9 +26,22 @@ namespace emp {
...
+        size_t data = Seed;
+        for (const auto& x : v) {
```
Comment (mmore500) [inline source/tools/hash_utils.h:37]: item instead of x

File: source/tools/hash_utils.h [hunk 3]
```diff
@@ -24,9 +26,22 @@ namespace emp {
...
+            using T = typename std::decay<decltype(x)>::type;
```
Comment (mmore500) [inline source/tools/hash_utils.h:38]: more descriptive than T?

File: source/tools/hash_utils.h (hunk 2 above)
Comment (rodsan0) [reply source/tools/hash_utils.h]: done in 7796005

File: source/tools/hash_utils.h (hunk 3 above)
Comment (rodsan0) [reply source/tools/hash_utils.h]: in 7796005, changed to `item_type` to correspond with `item` from previous review

File: source/tools/hash_utils.h (hunk 1 above)
Comment (rodsan0) [reply source/tools/hash_utils.h:139]: added in 7796005

File: source/tools/matchbin_metrics.h [hunk 4]
```diff
@@ -37,6 +37,16 @@
 #include "../tools/hash_utils.h"
 #include "../tools/tuple_utils.h"
 
+#include "../../third-party/span-lite/include/nonstd/span.hpp"
```
[the remark this answers is not in the document]
Comment (mmore500) [reply source/tools/matchbin_metrics.h:40]: Maybe we should organize it as "source/polyfill/span.h"

File: tests/tools/BitSet.cc [hunk 5]
```diff
@@ -379,6 +410,7 @@ TEST_CASE("Test BitSet", "[tools]")
 	test_flip();
 	test_bit();
 	test_byte();
+  test_bytes();
```
Comment (mmore500) [inline tests/tools/BitSet.cc:413]: indentation?

File: tests/tools/hash_utils.cc [hunk 6]
```diff
@@ -8,8 +8,10 @@
 #include <sstream>
 #include <iostream>
 #include <unordered_set>
+#include <array>
+#include <unordered_map>
 
-// // TODO: add asserts
+// TODO: add asserts
```
Comment (mmore500) [inline tests/tools/hash_utils.cc:14]: on what? (also, if it takes twenty or less minutes we should just do it before merging)

File: source/tools/BitSet.h [hunk 7]
```diff
@@ -589,6 +597,14 @@ namespace emp {
       return (bit_set[field_id] >> pos_id) & 255;
     }
 
+    /// Get a read-only view into the internal array used by BitSet
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/BitSet.h:600]: the rest of the file doesn't do this. should I still add it?

[the remark this answers is not in the document]
Comment (mmore500) [reply source/tools/BitSet.h:600]: Yeah :/ we should at least add it where we put in new code

File: source/tools/matchbin_metrics.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/matchbin_metrics.h]: moved into source/polyfill/span.h

File: source/tools/hash_utils.h [hunk 8]
```diff
@@ -24,9 +36,139 @@ namespace emp {
...
+      k ^= k >> 33;
+      k *= 0xc4ceb9fe1a85ec53;
+      k ^= k >> 33;
+    }
+  }
+
+  // implementation of the murmur3 hash, a fast hash with low collisions
+  // this hash makes it suitable for hash-based lookups
+  // for more info, see: https://en.wikipedia.org/wiki/MurmurHash
+  // this implementation was directly based on: https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/hash_utils.h]: done

File: source/tools/BitSet.h [hunk 9]
```diff
@@ -32,7 +32,15 @@
 #include "hash_utils.h"
 #include "random_utils.h"
 
-
+#include "../../third-party/span-lite/include/nonstd/span.hpp"
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/BitSet.h]: done!

File: source/tools/matchbin_metrics.h [hunk 10]
```diff
@@ -97,10 +99,13 @@ namespace emp {
...
-        qhasher(a),
-        thasher(b)
-      )) / (static_cast<double>(std::numeric_limits<size_t>::max()));
+      size_t input[2]{qhasher(a), thasher(b)};
+      const std::span<const std::byte> input_as_bytes(
+          reinterpret_cast<std::byte*>(input),
+          sizeof(input)
+      );
+      const size_t hash = emp::murmur_hash(input_as_bytes);
+      return static_cast<double>(hash) / (static_cast<double>(std::numeric_limits<size_t>::max()));
```
Comment (mmore500) [inline source/tools/matchbin_metrics.h:108]: delete extra parens

File: tests/tools/BitSet.cc [hunk 11]
```diff
@@ -144,6 +144,37 @@ void test_byte(){
...
+// actual testing function
+template <size_t Bits>
+void do_byte_test() {
+  emp::BitSet<Bits> bs;
+
+  for (size_t i = 0; i < Bits / 8; ++i) {
+    bs.SetByte(i, 10 * i);
+  }
+
+  auto myspan = bs.GetBytes();
```
Comment (mmore500) [inline tests/tools/BitSet.cc:159]: const?

File: tests/tools/BitSet.cc (hunk 5 above)
Comment (mmore500) [inline tests/tools/BitSet.cc:413]: Is this gh rendering the indentation funky?

File: tests/tools/hash_utils.cc (hunk 6 above)
Comment (mmore500) [reply tests/tools/hash_utils.cc:14]: lol this is my old comment let’s delete this

File: source/tools/hash_utils.h [hunk 12]
```diff
@@ -1,32 +1,178 @@
...
   }
 
+  /// This structure serves as a hash for containers that are iterable.
+  /// Use as a drop-in replacement for std::hash.
+  template <typename Container, size_t Seed = 0>
+  struct ContainerHash
+  {
+    size_t operator()(const Container& v) const {
+        size_t data = Seed;
+        for (const auto& item : v) {
```
Comment (mmore500) [inline source/tools/hash_utils.h:168]: Actually, could we use std accumulate here?

**Pull request merged**